### PR TITLE
enforce a stable list order when attaching/detaching instance groups

### PR DESCRIPTION
**Pull request by ryanpetrello**
related: https://github.com/ansible/awx/issues/3840

### CODE REVIEW

**Comment by ryanpetrello** [comment]
Prior to this PR, the M2M relations are returned in order based on `InstanceGroup.id` (based on insertion order):

---

```sql
SELECT main_instancegroup.id, main_inventory_instance_groups.inventory_id, name FROM main_inventory_instance_groups, main_instancegroup WHERE main_inventory_instance_groups.instancegroup_id = main_instancegroup.id;
 id | inventory_id | name
----+--------------+------
  2 |            1 | ig1
  4 |            1 | ig3
  3 |            1 | ig2
```
```python
>>> [ig.name for ig in Inventory.objects.get(pk=1).instance_groups.all()]
['ig1', 'ig2', 'ig3']
```

---

```sql
SELECT main_instancegroup.id, main_organization_instance_groups.organization_id, name FROM main_organization_instance_groups, main_instancegroup WHERE main_organization_instance_groups.instancegroup_id = main_instancegroup.id;
 id | organization_id | name
----+-----------------+------
  4 |               1 | ig3
  3 |               1 | ig2
  2 |               1 | ig1
```
```python
>>> [ig.name for ig in Organization.objects.get(pk=1).instance_groups.all()]
['ig1', 'ig2', 'ig3']
```

---

```sql
SELECT main_instancegroup.id, main_unifiedjobtemplate_instance_groups.unifiedjobtemplate_id, name FROM main_unifiedjobtemplate_instance_groups, main_instancegroup WHERE main_unifiedjobtemplate_instance_groups.instancegroup_id = main_instancegroup.id;
 id | unifiedjobtemplate_id | name
----+-----------------------+------
  4 |                     5 | ig3
  3 |                     5 | ig2
  2 |                     5 | ig1
```
```python
>>> [ig.name for ig in JobTemplate.objects.get(pk=5).instance_groups.all()]
['ig1', 'ig2', 'ig3']
```

**Comment by ryanpetrello** [comment]
Post-migration:

```sql
SELECT main_inventoryinstancegroupmembership.*, name FROM main_inventoryinstancegroupmembership, main_instancegroup WHERE main_inventoryinstancegroupmembership.instancegroup_id=main_instancegroup.id;
 id | position | instancegroup_id | inventory_id | name
----+----------+------------------+--------------+------
  1 |        0 |                2 |            1 | ig1
  2 |        1 |                3 |            1 | ig2
  3 |        2 |                4 |            1 | ig3

SELECT main_organizationinstancegroupmembership.*, name FROM main_organizationinstancegroupmembership, main_instancegroup WHERE main_organizationinstancegroupmembership.instancegroup_id=main_instancegroup.id;
 id | position | instancegroup_id | organization_id | name
----+----------+------------------+-----------------+------
  1 |        0 |                2 |               1 | ig1
  2 |        1 |                3 |               1 | ig2
  3 |        2 |                4 |               1 | ig3

SELECT main_unifiedjobtemplateinstancegroupmembership.*, name FROM main_unifiedjobtemplateinstancegroupmembership, main_instancegroup WHERE main_unifiedjobtemplateinstancegroupmembership.instancegroup_id=main_instancegroup.id;
 id | position | instancegroup_id | unifiedjobtemplate_id | name
----+----------+------------------+-----------------------+------
  1 |        0 |                2 |                     5 | ig1
  2 |        1 |                3 |                     5 | ig2
  3 |        2 |                4 |                     5 | ig3
```

File: awx/main/models/ha.py
Comment (ryanpetrello) [inline awx/main/models/ha.py]: context: https://github.com/ansible/awx/issues/3840#issuecomment-489757836

**Comment by AlanCoding** [comment]
I'm going to be honest that I dislike addressing this in the model / field classes. It feels like you are fighting the Django ORM. Managing your own Django field types is hard (migration implications, stale app definitions), we have had a long history of migration problems in fallout from the implicit role fields, which are still poorly understood (by me at least) to this day. Creating a new variant of M2M fields strikes me as similarly risky. We're also about to upgrade to Django 2.0.

Here's a very rudimentary sketch-up of how I would prefer to solve it in the view classes instead. This assumes that the field is defined as a M2M with the through model specified.

```diff
diff --git a/awx/api/generics.py b/awx/api/generics.py
index 2f2eca791b..ea5fa9b0da 100644
--- a/awx/api/generics.py
+++ b/awx/api/generics.py
@@ -618,6 +618,23 @@ class SubListCreateAPIView(SubListAPIView, ListCreateAPIView):
         return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
 
 
+def through_model_fields_names(model, relationship):
+    '''
+    Given a Django model or object for 'model',
+    this returns a 2-tuple of the field names for its through model,
+    where the first element is the reference back to the original model,
+    and the second element is the related model via the many-to-many field
+    '''
+    manager = getattr(model, relationship)
+    field = model._meta.get_field(relationship)
+    ret = (field.m2m_reverse_field_name(), field.m2m_field_name())
+    if manager.reverse:
+        # if we got to the field from the other model, then we must swap
+        return reversed(ret)
+    return ret
+
+
 class SubListCreateAttachDetachAPIView(SubListCreateAPIView):
     # Base class for a sublist view that allows for creating subobjects and
     # attaching/detaching them from the parent.
@@ -640,6 +657,13 @@ class SubListCreateAttachDetachAPIView(SubListCreateAPIView):
             res = Response(data, status=status.HTTP_400_BAD_REQUEST)
         return (sub_id, res)
 
+    def perform_attach(self, parent, sub):
+        parent_field_name, sub_field_name = through_model_fields_names(self.parent_model, self.relationship)
+        getattr(self.parent_model, relationship).through_model.create(**{
+            parent_field_name: parent,
+            sub_field_name: sub
+        })
+
     def attach(self, request, *args, **kwargs):
         created = False
         parent = self.get_parent_object()
@@ -681,7 +705,7 @@ class SubListCreateAttachDetachAPIView(SubListCreateAPIView):
 
         # Attach the object to the collection.
         if sub not in relationship.all():
-            relationship.add(sub)
+            self.perform_attach(parent, sub)
 
         if created:
             headers = {}
```

I know there's a lot to dislike here, but when things go wrong, I know we'll have vanilla M2M managers. Then we'll only be having to fix the use of the managers, not fixing the managers themselves.

The relevant subclasses would probably make some targeted modification to `perform_attach`.

I'm still reading through the full diff in this PR here.

**Comment by ryanpetrello** [comment]
@AlanCoding,

> I'm going to be honest that I dislike addressing this in the model / field classes. It feels like you are fighting the Django ORM. Managing your own Django field types is hard (migration implications, stale app definitions), we have had a long history of migration problems in fallout from the implicit role fields, which are still poorly understood (by me at least) to this day. Creating a new variant of M2M fields strikes me as similarly risky. We're also about to upgrade to Django 2.0.

I 100% agree.  I think this diff is gross, and getting here was extremely hard-fought (and probably will be again when we upgrade Django).  The Django ORM does **not** make this easy.

I think this _would_ be better served by special-casing the view class as you suggested.  When we upgrade to a version of Django that _does_ support these operations on M2M + `through=`, maybe we can just remove the special case code.

That said, removing our usage of `.add()` and `.remote()` doesn't actually win us much - it *only* gives us the ability to stop using the `auto_created` hack.  The painful thing about all this is that - as best as I can tell, we *do* need to use a custom field + descriptor to actually *apply* the default sort order (at least, I can't find _any_ other code or mechanism in the 1.11 implementation that provides for it).

**Comment by AlanCoding** [comment]
> as best as I can tell, we do need to use a custom field + descriptor to actually apply the default sort order (at least, I can't find any other code or mechanism in the 1.11 implementation that provides for it).

Yikes, I do see better where you're coming from on this.

there really wouldn't be any good way in the view to do anything about the default ordering. This is quite a pickle.

**Comment by ryanpetrello** [comment]
@AlanCoding another way we could potentially solve this would be to create a series of *database* views (in postgres) that applied an `ORDER_BY=position` condition, and we could have models in AWX that referenced those tables with `Meta.managed = False`.

That's probably also pretty cumbersome, though.

File: awx/main/fields.py
Comment (ryanpetrello) [inline awx/main/fields.py:1021]: It looks like the `django-sortedm2m` plugin does something _very_ similar to this:

https://github.com/gregmuellegger/django-sortedm2m/blob/master/sortedm2m/fields.py#L42

File: awx/main/tests/functional/test_instance_group_ordering.py
Comment (AlanCoding) [inline awx/main/tests/functional/test_instance_group_ordering.py:36]: nice vividly descriptive use of assert

File: awx/main/models/ha.py
Comment (AlanCoding) [inline awx/main/models/ha.py]: We could probably obtain this from the field to obliviate the need for this parameter, probably `.m2m_reverse_field_name()` or something. But I'm sure we can get by without that.

Comment (AlanCoding) [reply awx/main/models/ha.py]: Or `cls.lower()` as you did in the migration, assuming you could find the "from" model class

No wait, that won't work because of UJT vs JT, etc.

File: awx/main/migrations/0076_v360_add_new_instance_group_relations.py
Comment (AlanCoding) [inline awx/main/migrations/0076_v360_add_new_instance_group_relations.py:6]: Question:

Let me play out the future in my head.

We merge this
Someone starts upgrading to Django 2.0
On startup errors are thrown, the add and remove methods are deleted from `OrderedManyToManyDescriptor`
then `OrderedManyToManyDescriptor` may or may not remain necessary, depending on if Django will let us order the through model

Do we still keep `OrderedManyToManyField`?

We're importing the field here, so if we deleted the field class, then we would have to change the migration.

But it looks more like, for the Django 2.0 upgrade, we just delete some of the most painful parts of the hack, but the ordered M2M field will remain a thing.

**Review by AlanCoding — APPROVED** [review]

File: awx/main/migrations/0076_v360_add_new_instance_group_relations.py
Comment (ryanpetrello) [reply awx/main/migrations/0076_v360_add_new_instance_group_relations.py:6]: Yep, my thought is we keep this ordered field - even in Django 2, we still need it to apply the default sort order, and we'll need a custom descriptor to do so.

At upgrade time, we just lose the custom `.add()` and `.remove()` methods.

Comment (ryanpetrello) [reply awx/main/migrations/0076_v360_add_new_instance_group_relations.py:6]: cc @jbradberry

File: awx/ui/client/src/instance-groups/instance-groups.service.js
Comment (jakemcdermott) [inline awx/ui/client/src/instance-groups/instance-groups.service.js]: compare the currently saved ids and the selected ids - as soon as we encounter a difference between the two arrays, we mark _all_ remaining currently saved ids for deletion.

Then, we build the association array in the correct order by adding any id that was marked for disassociation due to being out of order _and_ any ids that are new associations.

```
current:  [1,2,3,4,5,6]
selected: [1,2,4,5,6,3] 

-> idsToDisassociate = [3,4,5,6]
-> idsToAssociate = [2,3]
```

File: awx/ui/client/src/instance-groups/instance-groups.service.js
Comment (jakemcdermott) [inline awx/ui/client/src/instance-groups/instance-groups.service.js]: Now that I look at this, I'm not sure _all_ is safe to use here. I think we need to make each and every association / disassociation request sequentially...

File: awx/ui/client/src/instance-groups/instance-groups.service.js
Comment (jakemcdermott) [inline awx/ui/client/src/instance-groups/instance-groups.service.js]: Now that I look at this some more, I'm not sure _all_ is safe to use for the associations. I think we need to make each and every association  request sequentially.

File: awx/main/models/ha.py
Comment (ryanpetrello) [reply awx/main/models/ha.py]: Uploaded a solution to this.

File: awx/main/fields.py
Comment (ryanpetrello) [reply awx/main/fields.py:1021]: @AlanCoding @jbradberry this seems like reasonable behavior for *Django* to have.  I might put together a PR upstream and see if there's any interest (it would be an easy PR).

Maybe we could land this upstream and get rid of this custom descriptor in the future.

**Comment by chrismeyersfsu** [comment]
@ryanpetrello Doe this break the `order_by` GET param for endpoints that expose this model?

**Comment by chrismeyersfsu** [comment]
@ryanpetrello You need a test where you create a hole in the ordering i.e.
`group-1`
`group-2`
`group-3`
`group-4`

Delete `group-3`. Verify that all the positions shift.

File: awx/ui/client/src/instance-groups/instance-groups.service.js
Comment (jakemcdermott) [inline awx/ui/client/src/instance-groups/instance-groups.service.js:67]: we need to do the association and disassociation requests sequentially to make sure they're processed in the right order on the backend so we build a promise chain instead of using `.all`.

File: awx/ui/client/src/templates/job_templates/edit-job-template/job-template-edit.controller.js
Comment (jakemcdermott) [inline awx/ui/client/src/templates/job_templates/edit-job-template/job-template-edit.controller.js:612]: This bug already existed, but I started running into it a lot when testing the changes to the IG service. We need to add the IG (dis)association promises to the list of defers we wait for on save before reloading the form. Otherwise, we will sometimes reload the page before all of the association requests have completed. This makes it seem like some of the associations weren't made when they actually were.

**Comment by ryanpetrello** [comment]
@chrismeyersfsu 

Good feedback.  Confirmed they work:

https://github.com/ansible/awx/pull/3842/files#diff-93b4e811fc5321696b6e414f009fba68R60
https://github.com/ansible/awx/pull/3842/files#diff-93b4e811fc5321696b6e414f009fba68R88

image

```shell
curl -sk "https://ryan:ryan@awx:8043/api/v2/inventories/1/instance_groups/" | python -m json.tool | grep name
            "name": "2",
            "name": "3",
            "name": "tower",
            "name": "1",
```

```shell
curl -sk "https://ryan:ryan@awx:8043/api/v2/inventories/1/instance_groups/?order_by=name" | python -m json.tool | grep name
            "name": "1",
            "name": "2",
            "name": "3",
            "name": "tower",
```

```shell
curl -sk "https://ryan:ryan@awx:8043/api/v2/instance_groups/" | python -m json.tool | grep name
            "name": "tower",
            "name": "1",
            "name": "2",
            "name": "3",
```

```shell
curl -sk "https://ryan:ryan@awx:8043/api/v2/instance_groups/?order_by=name" | python -m json.tool | grep name
            "name": "1",
            "name": "2",
            "name": "3",
            "name": "tower",
```

File: awx/main/fields.py
Comment (jbradberry) [inline awx/main/fields.py]: I think that this `.object_name.lower()` is not necessary and may be fragile if we want to reuse or change things. The first portion of this path should be available as an attribute on the Field instance, I just need to remember which is relevant here.

File: awx/main/fields.py
Comment (jbradberry) [inline awx/main/fields.py]: Do we want to check against `.startswith('1.')` instead, just in case we wind up bumping the version before doing the full Django 2 upgrade?

Comment (ryanpetrello) [reply awx/main/fields.py]: I like this.

File: awx/main/migrations/0076_v360_add_new_instance_group_relations.py
Comment (jbradberry) [reply awx/main/migrations/0076_v360_add_new_instance_group_relations.py:6]: I have to admit, I don't love it.

Comment (jakemcdermott) [reply awx/main/migrations/0076_v360_add_new_instance_group_relations.py:6]: love is the answer

**Review by jbradberry — APPROVED** [review]
I think I'm as OK with this as I'm going to get.

**Comment by ryanpetrello** [comment]
I filed a bug in Django: https://code.djangoproject.com/ticket/30460

@AlanCoding @jbradberry here's a PR to Django that I think should give us what we want (if we can get something like this in a future Django 🤞):
https://github.com/django/django/pull/11340

**Comment by AlanCoding** [comment]
I replied to a question at:

https://github.com/gregmuellegger/django-sortedm2m/issues/128

this brings up a good question for our purposes, will `Organization.objects.prefetch_related('instance_groups')` work correctly, do we need to apply a custom prefetch in such cases, and does it matter (do such cases exist anyway)?

...just asking question for now, not answering them.

**Comment by ryanpetrello** [comment]
@AlanCoding 

```python
In [1]: from django.db import connection
In [2]: for o in Organization.objects.prefetch_related('instance_groups').all():
   ...:     for ig in o.instance_groups.all():
   ...:         print(ig.name)
   ...:
In [2]: for o in Organization.objects.prefetch_related('instance_groups').all():
   ...:     for ig in o.instance_groups.all():
   ...:         print(ig.name)
   ...:
3
2
1
tower
```

```sql
In [3]: connection.queries
Out[3]:
[{'sql': 'SELECT "main_organization"."id", "main_organization"."created", "main_organization"."modified", "main_organization"."description", "main_organization"."created_by_id", "main_organization"."modified_by_id", "main_organization"."name", "main_organization"."custom_virtualenv", "main_organization"."max_hosts", "main_organization"."admin_role_id", "main_organization"."execute_role_id", "main_organization"."project_admin_role_id", "main_organization"."inventory_admin_role_id", "main_organization"."credential_admin_role_id", "main_organization"."workflow_admin_role_id", "main_organization"."notification_admin_role_id", "main_organization"."job_template_admin_role_id", "main_organization"."auditor_role_id", "main_organization"."member_role_id", "main_organization"."read_role_id" FROM "main_organization" ORDER BY "main_organization"."name" ASC',
  'time': '0.003'},
 {'sql': 'SELECT ("main_organizationinstancegroupmembership"."organization_id") AS "_prefetch_related_val_organization_id", "main_instancegroup"."id", "main_instancegroup"."name", "main_instancegroup"."created", "main_instancegroup"."modified", "main_instancegroup"."controller_id", "main_instancegroup"."policy_instance_percentage", "main_instancegroup"."policy_instance_minimum", "main_instancegroup"."policy_instance_list" FROM "main_instancegroup" INNER JOIN "main_organizationinstancegroupmembership" ON ("main_instancegroup"."id" = "main_organizationinstancegroupmembership"."instancegroup_id") WHERE "main_organizationinstancegroupmembership"."organization_id" IN (1)',
  'time': '0.001'},
 {'sql': 'SELECT "main_instancegroup"."id", "main_instancegroup"."name", "main_instancegroup"."created", "main_instancegroup"."modified", "main_instancegroup"."controller_id", "main_instancegroup"."policy_instance_percentage", "main_instancegroup"."policy_instance_minimum", "main_instancegroup"."policy_instance_list" FROM "main_instancegroup" INNER JOIN "main_organizationinstancegroupmembership" ON ("main_instancegroup"."id" = "main_organizationinstancegroupmembership"."instancegroup_id") WHERE "main_organizationinstancegroupmembership"."organization_id" = 1 ORDER BY "main_organizationinstancegroupmembership"."position" ASC',
  'time': '0.001'}]
```

**Comment by ryanpetrello** [comment]
@AlanCoding it seems like _more_ queries than not specifying `prefetch_related`.  Am I doing this wrong?

```python
In [1]: from django.db import connection
In [2]: for o in Organization.objects.prefetch_related('instance_groups').all():
   ...:     for ig in o.instance_groups.all():
   ...:         print(ig.name)
   ...:
3
2
1
tower
```

```sql
In [3]: connection.queries
Out[3]:
[{'sql': 'SELECT "main_organization"."id", "main_organization"."created", "main_organization"."modified", "main_organization"."description", "main_organization"."created_by_id", "main_organization"."modified_by_id", "main_organization"."name", "main_organization"."custom_virtualenv", "main_organization"."max_hosts", "main_organization"."admin_role_id", "main_organization"."execute_role_id", "main_organization"."project_admin_role_id", "main_organization"."inventory_admin_role_id", "main_organization"."credential_admin_role_id", "main_organization"."workflow_admin_role_id", "main_organization"."notification_admin_role_id", "main_organization"."job_template_admin_role_id", "main_organization"."auditor_role_id", "main_organization"."member_role_id", "main_organization"."read_role_id" FROM "main_organization" ORDER BY "main_organization"."name" ASC',
  'time': '0.003'},
 {'sql': 'SELECT ("main_organizationinstancegroupmembership"."organization_id") AS "_prefetch_related_val_organization_id", "main_instancegroup"."id", "main_instancegroup"."name", "main_instancegroup"."created", "main_instancegroup"."modified", "main_instancegroup"."controller_id", "main_instancegroup"."policy_instance_percentage", "main_instancegroup"."policy_instance_minimum", "main_instancegroup"."policy_instance_list" FROM "main_instancegroup" INNER JOIN "main_organizationinstancegroupmembership" ON ("main_instancegroup"."id" = "main_organizationinstancegroupmembership"."instancegroup_id") WHERE "main_organizationinstancegroupmembership"."organization_id" IN (1)',
  'time': '0.001'},
 {'sql': 'SELECT "main_instancegroup"."id", "main_instancegroup"."name", "main_instancegroup"."created", "main_instancegroup"."modified", "main_instancegroup"."controller_id", "main_instancegroup"."policy_instance_percentage", "main_instancegroup"."policy_instance_minimum", "main_instancegroup"."policy_instance_list" FROM "main_instancegroup" INNER JOIN "main_organizationinstancegroupmembership" ON ("main_instancegroup"."id" = "main_organizationinstancegroupmembership"."instancegroup_id") WHERE "main_organizationinstancegroupmembership"."organization_id" = 1 ORDER BY "main_organizationinstancegroupmembership"."position" ASC',
  'time': '0.001'}]
```

**Review by marshmalien — APPROVED** [review]
Nice work @jakemcdermott. By adding comments and spelling out each step the instance groups service code is much more readable. The UI portion has my stamp of approval.

File: awx/ui/client/src/templates/job_templates/edit-job-template/job-template-edit.controller.js
Comment (marshmalien) [reply awx/ui/client/src/templates/job_templates/edit-job-template/job-template-edit.controller.js:612]: 👍Nice

**Comment by ryanpetrello** [comment]
@one-t @unlikelyzero @kdelee @elyezer 

Any of you have some time to give this a whirl before merging?  I'd _especially_ like to test out migration/upgrades to make sure I got this right.

File: awx/ui/client/src/instance-groups/instance-groups.service.js
Comment (jakemcdermott) [inline awx/ui/client/src/instance-groups/instance-groups.service.js:22]: This function compares the currently saved ids and the selected ids - as soon as we encounter a difference between the two arrays, we mark _all_ remaining currently saved ids in the array for disassociation.

Then, we build the association array in the correct order by adding any selected id that was marked for disassociation due to being out of order _and_ any ids that are new associations.

```
# scenario:
# -----------
# page is loaded with [1,2,3,4,5,6] as the currently selected tags
# user removes tag 3 from the middle
# user adds a new tag 7 to the end
# user appends tag 3 to the end

                ___________ all ids here and to the right are disassociated
               |
current:  [1,2,3,4,5,6]
selected: [1,2,4,5,6,7,3]
               |___________ all ids here and to the right are (re)associated 

-> idsToDisassociate = [3,4,5,6]
-> idsToAssociate = [4,5,6,7,3]
```

When sending requests to the API, we do all of the disassociations first (and wait for their completion) before doing the associations.

**Comment by kdelee** [comment]
trying this out now

**Review by kdelee — APPROVED** [review]
:+1: Works for me!

Also have to say that editing both in the small window on normal UI + the modal works seamlessly. thanks

**Pull request merged**
